File: CrossfitDiary/CoreApp/CrossfitDiaryCore.Web/ClientApp/models/viewModels/ExerciseViewModel.ts
import Deserializable = General.Deserializable;
import {ExerciseMeasureViewModel} from "./ExerciseMeasureViewModel";
import {ExerciseMeasureType} from "./ExerciseMeasureType";
import {WeightDisplayType} from "./WeightDisplayType";

export class ExerciseViewModel implements Deserializable {
    id: number = 0;
    title?: string;
    abbreviation?: string;
    exerciseMeasures: ExerciseMeasureViewModel[] = [];
    isAlternative: boolean = false;
    isNewWeightMaximum?: boolean = false;
    isDoUnbroken: boolean = false;
    addedToMaxWeightString?: string;
    weightDisplayType: WeightDisplayType = WeightDisplayType.Default;
    count?: string = null;
    weight?: string = null;
    calories?: string = null;
    weightPercentValue?: number = null;
    
    countMeasure: ExerciseMeasureViewModel;
    weightMeasure: ExerciseMeasureViewModel;
    altWeightMeasure: ExerciseMeasureViewModel;
    caloriesMeasure: ExerciseMeasureViewModel;
    distanceMeasure: ExerciseMeasureViewModel;
    heightMeasure: ExerciseMeasureViewModel;
    timeMeasure: ExerciseMeasureViewModel;


    calculatedWeight?: number = null;

    constructor(input?: any) {
        if (input == null) {
            return;
        }
        Object.assign(this, input);
    }

    public deserialize(input: any): this {
        if (input == null) {
            return;
        }
        Object.assign(this, input);

        this.exerciseMeasures = input.exerciseMeasures.map(x => new ExerciseMeasureViewModel().deserialize(x));

        this.count = this.getMeasureValue(ExerciseMeasureType.Count);
        this.weight = this.getMeasureValue(ExerciseMeasureType.Weight);
        this.calories = this.getMeasureValue(ExerciseMeasureType.Calories);
        
        this.countMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.Count);
        this.weightMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.Weight);
        this.altWeightMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.AlternativeWeight);
        this.caloriesMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.Calories);
        this.distanceMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.Distance);
        this.heightMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.Height);
        this.timeMeasure = this.exerciseMeasures.find(x => x.measureType === ExerciseMeasureType.Time);
        
        return this;
    }

    public haveSameCountAndCalories = (toCompareExercise: ExerciseViewModel): boolean => {
        if (this.count == null && toCompareExercise.count == null) {
            if (this.calories == null && toCompareExercise.calories == null) {
                return false;
            } else {
                return this.calories === toCompareExercise.calories;
            }
        } else {
            if (this.calories == null && toCompareExercise.calories == null) {
                return this.count === toCompareExercise.count;
            } else {
                let result: boolean =
                    this.count === toCompareExercise.count ||
                    this.count === toCompareExercise.calories ||
                    this.calories === toCompareExercise.count ||
                    this.calories === toCompareExercise.calories;
                return result;
            }
        }
    };

    getMeasureValue(measureType: ExerciseMeasureType): string {
        let foundCountMeasure: ExerciseMeasureViewModel = this.exerciseMeasures.find(x => x.measureType === measureType);
        let result: string = foundCountMeasure ? foundCountMeasure.measureValue : null;
        return result;
    }

    resetWeightMeasures() {
        this.weightMeasure.measureValue = null;
        this.altWeightMeasure.measureValue = null;
        this.weightDisplayType = WeightDisplayType.Default;
        this.weightPercentValue = null;
    }
}

File: CrossfitDiary/CoreApp/CrossfitDiaryCore.Web/ViewModels/ExerciseViewModel.cs
using System.Collections.Generic;
using Newtonsoft.Json;

namespace CrossfitDiaryCore.Web.ViewModels
{
    public class ExerciseViewModel
    {
        [JsonProperty("id")]
        public int Id { get; set; }

        [JsonProperty("title")]
        public string Title { get; set; } 
        
        [JsonProperty("abbreviation")]
        public string Abbreviation { get; set; }

        [JsonProperty("exerciseMeasures")]
        public IEnumerable<ExerciseMeasureViewModel> ExerciseMeasures { get; set; }

        [JsonProperty("isAlternative")]
        public bool IsAlternative { get; set; }

        [JsonProperty("isNewWeightMaximum")]
        public bool IsNewWeightMaximum { get; set; }

        [JsonProperty("isDoUnbroken")]
        public bool IsDoUnbroken { get; set; }

        [JsonProperty("addedToMaxWeightString")]
        public string AddedToMaxWeightString { get; set; }

        [JsonProperty("position")]
        public int Position { get; set; }

        public int? TempRoutineSimpleId { get; set; }

        [JsonProperty("weightDisplayType")]
        public WeightDisplayType WeightDisplayType { get; set; } = WeightDisplayType.PercentMaxPM;

    }


    public class ExerciseMeasureViewModel
    {
        [JsonProperty("measureType")]
        public MeasureTypeViewModel MeasureType { get; set; }

        [JsonProperty("measureValue")]
        public string MeasureValue { get; set; }

        [JsonProperty("description")]
        public string Description { get; set; }

        [JsonProperty("shortMeasureDescription")]
        public string ShortMeasureDescription { get; set; }

        [JsonProperty("isRequired")]
        public bool IsRequired { get; set; } = true;


    }

    public enum WeightDisplayType
    {
        Default = 0,
        PercentPreviousPM,
        PercentMaxPM = 2
    }


    public enum MeasureTypeViewModel
    {
        Distance = 1,
        Count = 2,
        Weight = 3,
        Calories = 4,
        Height = 5,
        AlternativeDistance = 6,
        AlternativeCount = 7,
        AlternativeWeight = 8,
        AlternativeCalories = 9,
        AlternativeHeight = 10,
    }
}